pg-extend/build.rs

```diff
@@ -63,6 +62,95 @@ fn main() {
     println!("cargo:rustc-cfg=feature=\"{}\"", feature_version);
 }
 
+#[cfg(windows)]
+fn get_bindings(pg_include: &str) -> bindgen::Builder {
+    // Compilation in windows requires these extra inclde paths
+    let pg_include_win32_msvc = format!("{}\\port\\win32_msvc", pg_include);
+    let pg_include_win32 = format!("{}\\port\\win32", pg_include);
+    // The `pg_include` path comes in the format og "includes/server", but we also need
+    // the parent folder, so we remove the "/server" part at the end
+    let pg_include_parent = pg_include[..(pg_include.len() - 7)].to_owned();
+
+    bindgen::Builder::default()
+        .clang_arg(format!("-I{}", pg_include_win32_msvc))
+        .clang_arg(format!("-I{}", pg_include_win32))
+        .clang_arg(format!("-I{}", pg_include))
+        .clang_arg(format!("-I{}", pg_include_parent))
+        // Whitelist all PG-related functions
+        .whitelist_function("pg.*")
+        // Whitelist used functions
+        .whitelist_function("longjmp")
+        .whitelist_function("_setjmp")
+        .whitelist_function("cstring_to_text")
+        .whitelist_function("text_to_cstring")
+        .whitelist_function("errmsg")
+        .whitelist_function("errstart")
+        .whitelist_function("errfinish")
+        .whitelist_function("pfree")
+        // Whitelist all PG-related types
```

Comment:
This whitelist can limit the usefulness of `pg-extend-rs`; some extension modules use `pg_sys` to get access to additional PostgreSQL functions without needing to run bindgen themself.

Have you considered any alternatives to such whitelisting?

Reply:
I agree it can limit the usefulness, but it generates a HUGE file if you leave as default everything:
- Linux generated file is 38k LoC
- Windows generated file is 132k LoC

I'm not confortable with having a hard-coded list, that's why I limited only for Windows while an alternative is figured out

Reply:
Could we introduce a feature flag that would enable all symbols vs. limit them?

“all-symbols” seems fairly obvious. How we dice and slice the smaller sets based on which features you’re enabling is a little harder to see.

Reply:
The problem I encountered was that "allowing all" introuced a lot of things (if I'm not wrong, it imports the full Windows API) which I started by blacklisting them, but the list is so big, that it's simpler to do a whitelisting (which works recursively) of what was used from the crate.

As I mentioned, this whitelisting based approach is only for Windows. I left the "include all" approach for the `unix` family

Reply:
I'm open to suggestions on how to tackle the white/black listing. One option would be to be able to pull the list of all the PG symbols and use it for the whitelisting, but I have no idea on how to do that or where to pull it from. If understand correctly how `bindgen` works, whitelisting that list of symbols will recursively include any non-PG symbols that the PG ones use

Reply:
Even if it does include the entirety of Windows API headers, what problems does that cause? Looking at bindgen's issue tracker, it seems that's a common way to use it. (e.g. https://github.com/rust-lang/rust-bindgen/issues/1562 "I'm pretty sure on Firefox bindgen ends up including windows.h without issues")

You said it's slow to compile; is that the only reason why you want whitelisting? How much time does it save?

If it's merely an optimization of build speed, I don't see why this needs to be Windows-specific at all. Omitting potentially useful symbols from a module depending on platform, for only build speed reasons, seems like very surprising behavior to me, likely to cause porting issues to downstream users.

If the speed gains are worthwhile, a feature flag, as suggested by @bluejekyll, sounds like a great idea. Then Unix builds could also take advantage of it, if desired. But that would better be done as a separate PR IMO, and strip whitelisting from this PR.

Reply:
Thanks for the comment @intgr. The reason to not include all was not due to slow compilation, but because it includes problematic code that has issues at the Rust compiler level.
```sh
error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
     --> E:\rust-projects\pg-extend-rs\target\release\build\pg-extend-7be983388aa96263\out/postgres.rs:32012:1
      |
32012 | / pub struct _IMAGE_TLS_DIRECTORY64 {
32013 | |     pub StartAddressOfRawData: ULONGLONG,
32014 | |     pub EndAddressOfRawData: ULONGLONG,
32015 | |     pub AddressOfIndex: ULONGLONG,
...     |
32018 | |     pub __bindgen_anon_1: _IMAGE_TLS_DIRECTORY64__bindgen_ty_1,
32019 | | }
      | |_^

error[E0277]: arrays only have std trait implementations for lengths 0..=32
     --> E:\rust-projects\pg-extend-rs\target\release\build\pg-extend-7be983388aa96263\out/postgres.rs:51363:5
      |
51363 |     pub __bindgen_padding_0: [u8; 40usize],
      |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::array::LengthAtMost32` is not implemented for `[u8; 40]`
      |
      = note: required because of the requirements on the impl of `std::fmt::Debug` for `[u8; 40]`
      = note: required because of the requirements on the impl of `std::fmt::Debug` for `&[u8; 40]`
      = note: required for the cast to the object type `dyn std::fmt::Debug`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `pg-extend`.
```
I first tried the blacklisting approach but it was a difficult one so I moved to the whitelisting approach.

On the other hand (based on what I understood from @bluejekyll's comment, please correct me if I'm wrong), the list of whitelisted symbols could be based on feature flags, but I would still aim for the "include all" to be to whitelist all PG related things, which will recursivelly include any extra things that are required.

The fact that there's a ~130k LoC reduction was just a bonus, not the main reason for the Pr

Reply:
Looks like this issue
> error[E0277]: arrays only have std trait implementations for lengths 0..=32

would be solved by updating bindgen. Out `Cargo.lock` file references bindgen 0.48.1 which is more than half a year old!

The other issue might be solved as well (https://github.com/rust-lang/rust-bindgen/issues/1498?). Give it a shot.

Reply:
I bumped `bindgen` to 0.51.1 and still the same errors